### PR TITLE
[FIRRTL] Keep classes around

### PR DESCRIPTION
**Pull request by rwy7**
Uses of classes are not always discoverable, this PR makes two changes to prevent the firtool pass pipeline from removing classes that appear dead.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 1]
```diff
@@ -1631,6 +1631,13 @@ BlockArgument ClassOp::getArgument(size_t portNumber) {
   return getBodyBlock()->getArgument(portNumber);
 }
 
+bool ClassOp::canDiscardOnUseEmpty() {
+  // ClasseOps are referenced by ClassTypes, and these uses are not
+  // discovereable by the symbol infrastructure. Return false here to prevent
+  // passes like symbolDCE from removing our classes.
+  return false;
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1638]: This part makes sense to me.

**Review by mikeurbach — APPROVED** [review]
Less sure about inliner, but I think canDiscardOnUseEmpty makes sense for ClassOp. MLIR needs better infra for this (like a SymbolUserTypeInterface, or something).

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp [hunk 2]
```diff
@@ -1395,7 +1395,7 @@ void Inliner::run() {
 
   // Mark the top module as live, so it doesn't get deleted.
   for (auto module : circuit.getOps<FModuleLike>()) {
-    if (!module.isPublic())
+    if (!module.isPublic() && !isa<ClassOp>(module))
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:1398]: WDYT about:
```suggestion
    if (module.canDiscardOnUseEmpty())
```

The default implementation of that returns `!isPublic()` already.  Not sure about nested, probably fine to not worry for now.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 1 above)
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1638]: Looking a little, looks like SymbolDCE only walks attributes (`op->getAttrDictionary().walk<WalkOrder::PreOrder>([&](SymbolRefAttr symbolRef)``, is that the reason?

Sounds good to me, we can DCE these ourselves as-needed in a few places.

**Review by dtzSiFive — APPROVED** [review]

**Comment by dtzSiFive** [comment]
I want to suggest a test, but other than checking SymbolDCE (or our inliner) specifically not sure offhand what to suggest.  If there's something you think of great otherwise this seems fine as-is.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 1 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1638]: Yep.

So I think eventually we should have a FIRRTL syntax for marking classes as public/private, and then a class can be kept alive by either being public, or being instantiated by an object op. But until then, we might as well be conservative.

File: lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp (hunk 2 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp]: This is much much better, thanks!

**Pull request merged**